### PR TITLE
Improve clang and GCC diagnostics

### LINKED ISSUE #1511: Flycheck for C++
I have a file "framework.cpp"
with single line`#include "framework.h"`
`C-c ! l` shows next:

```
     1     warning         In include ./framework.h (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
     1     error           In file included from (c/c++-clang)
```
Switching to file `framework.h`
```
1   9 warning         #pragma once in main file (c/c++-clang)
     3     error           In include ./render/renderer.h (c/c++-clang)
     3     error           In file included from (c/c++-clang)
```
Switching to file `render/renderer.h`
```
1   9 warning         #pragma once in main file (c/c++-clang)
     3     warning         In include ./device/device.h (c/c++-clang)
     4     error           In file included from (c/c++-clang)
     4     error           In file included from (c/c++-clang)
     4     warning         In include ./draw_stages/draw_stages.h (c/c++-clang)
     4     error           In file included from (c/c++-clang)
     4     error           In file included from (c/c++-clang)
     4     error           In file included from (c/c++-clang)
     4     error           In file included from (c/c++-clang)
     4     error           In file included from (c/c++-clang)
     4     error           In file included from (c/c++-clang)
     4     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
     5     error           In file included from (c/c++-clang)
```
And etc.. etc... etc..
And finally in the some file
```
1   9 warning         #pragma once in main file (c/c++-clang)
     4  10 error           'cstring' file not found (c/c++-clang)
```
```
#pragma once

#include <vulkan/vulkan.h>
#include <cstring>

//more code
```
What I expected: no 100 deep file routines, no useless spam like one hundred "in file included from" without "PATH/FILE_THAT_INCLUDED", and finally `#include <cstring>` mustn't be an error.

`M-x flycheck-verify-setup`
```
Syntax checkers for buffer misc.h in c-mode:

  c/c++-clang
    - may enable: yes
    - executable: Found at /usr/lib/llvm/7/bin/clang

  c/c++-gcc
    - may enable: yes
    - executable: Found at /usr/bin/gcc

  c/c++-cppcheck
    - may enable: Automatically disabled!
    - executable: Not found

Flycheck Mode is enabled.  Use C-u C-c ! x to enable disabled
checkers.

--------------------

Flycheck version: 31
Emacs version:    26.1
System:           x86_64-pc-linux-gnu
Window system:    x
```
System `gentoo`

Also this config didn't work when I call `flycheck-compile c/c++-clang`, please tell me why.
```
;;; Directory Local Variables
;;; For more information see (info "(emacs) Directory Variables")

((c++mode .
	  (
	      (flycheck-clang-language-standard . c++17)
	      (flycheck-clacn-no-exceptions . t)
	      (flycheck-clang-no-rtti . t)
	      (flycheck-clang-warnings .  "everything no-pragma-once-outside-header no-missing-braces no-c++98-compat-pedantic no-nested-anon-types no-zero-length-array no-c++98-c++11-compat-binary-literal no-comma no-c++98-compat no-gnu-anonymous-struct no-c99-extensions no-vla-extension no-c++17-extensions no-switch-enum no-vla")
	      (flycheck-clang-args . nil)
	      )
))
```

### PR DESCRIPTION
**Pull request by fmdkdd**
Fix #1511.

~We may want to do the same for GCC as well.~  Done.

### CODE REVIEW

**Comment by fmdkdd** [comment]
One great thing would be to group together the `info` diagnostics stemming from an error.  Unfortunately, the current clang/gcc output is not structured, so at best we know that the lines "In file included from..." start a group, but when does a group end?  When another error appears maybe?  I don't feel good enough about this heuristic to use it.

File: flycheck.el [hunk 1]
```diff
@@ -651,6 +651,17 @@ using \\[flycheck-error-list-reset-filter]."
...
+(defcustom flycheck-relevant-error-other-file-min-level 'error
+  "The minimum level of errors from other files to display in this buffer.
+
+If set to an error level, only display errors from other files
+whose error level is at least as severe as this one list.  If
+nil, display all errors from other files."
+  :group 'flycheck
+  :type 'flycheck-minimum-level
+  :safe #'flycheck-error-level-p
+  :package-version '(flycheck . "32"))
```
Comment (cpitclaudel) [inline flycheck.el]: Should this be documented in the manual, too?

Comment (fmdkdd) [reply flycheck.el]: Good point.  I've added a section on "errors from other files" with this new variable.

**Comment by cpitclaudel** [comment]
In exciting recent news, gcc9 has json diagnostics, so we'll be able to get rid of our heuristics :)  Thanks @davidmalcolm!

**Pull request merged**